text/0000-pack-buildpack-new-alternatives.md

```diff
@@ -0,0 +1,210 @@
+
+`pack buildpack new --libcnb` supports end-users who wish to scaffold a new `libcnb`-based buildpack.  The operation of `pack buildpack new --libcnb` does not require internet access to succeed.  `pack buildpack new --template URL` allows projects (such as Paketo, Heroku and others) to define skeleton projects for new buildpacks.
+
+Extending `pack buildpack new` allows buildpack authors to more-easily create new buildpacks in their chosen language and framework.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Project scaffolding is [very](https://cookiecutter.readthedocs.io/) [popular](https://yeoman.io/) [in](https://github.com/kowainik/summoner) [other](https://crates.io/crates/cargo-scaffold) [ecosystems](https://github.com/golang-standards/project-layout).  Scaffolding systems are employed to ease onboarding of new developers.  Within `pack` this feature is targeted at onboarding buildpack authors.
+
+- Explaining the feature largely in terms of examples.
+
+`pack buildpack new --libcnb` should generate skeleton code according to the [golang standard project layout](https://github.com/golang-standards/project-layout) and include `libcnb` best practices.  The layout would look similar to the following:
+
+```bash
+.
+├── buildpack.toml
+├── cmd
+│   └── main.go
+├── go.mod
+├── go.sum
+└── pkg
+    ├── build.go
+    └── detect.go
+```
+
+This approach is not opinionated on topics such as which testing framework to use.  Only components under github.com/buildpacks/* are used in the `--libcnb` skeleton.  `pack buildpack new --template URL` allows for more opinionated project scaffolding.  As such, we support new buildpack authors with `--libcnb` onboarding and support experienced buildpack authors to use the scaffolding of their choice.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The design is modeled on [cookiecutter](https://cookiecutter.readthedocs.io/en/1.7.2/) with reference to [springerle](https://github.com/carlmjohnson/springerle) -- a similar implementation in golang -- and [create-go-app](https://github.com/create-go-app/).  We explicitly do not want to re-implement cookiecutter in golang due to the scope of such an implementation.  Nor do we want to `os.Exec` a python subprocess to run cookiecutter as this would require availability of a python runtime environment.  Instead we propose to borrow heavily from create-go-app, and generate the default shell and libcnb skeletons from an embedded file system.
+
+A full invocation to generate `libcnb` scaffolding is similar to the [documented project scaffolding](https://buildpacks.io/docs/buildpack-author-guide/create-buildpack/building-blocks-cnb/#using-the-pack-cli).
+
+```bash
+pack buildpack new --libcnb examples/ruby \
+    --api 0.7 \
```

Comment:
To me, `--api` implies that any template has to support all valid(?) buildpack API versions. Can a template decide to only support one API version? For example, [libcnb.rs ](https://github.com/Malax/libcnb.rs) currently only supports one API version.

If we offer `--api`, it would be nice if the template at least can warn that it does not support that API version, but it might even be worth letting the template have an opinion which API version is best and not offer `--api`.

Reply:
I've changed the proposal and now put this choice in the definition of the project template.